### PR TITLE
fix(session): drop interleaved tool rows after mid-history truncate (#5134)

### LINKED ISSUE #5134: fix(session): truncate context-alignment can over-keep interleaved tool rows on mid-history rewind (follow-up to #5096)
## Summary

Follow-up from the #5096 fix bundle (#5124 / #5126, shipped together). The shared context-truncation
helper `truncate_context_for_display_keep` (`api/session_ops.py`) aligns the model-facing
`context_messages` to the display `messages` **positionally**:

```python
prefix_len = max(0, len(ctx) - len(msgs))
return ctx[:prefix_len] + ctx[prefix_len:][:keep]
```

This is correct — and a strict improvement over the previous behavior — when the divergence between
`context_messages` and display `messages` is a **leading prefix** (e.g. compaction-only leading rows,
`len(ctx) > len(msgs)`). It fixes the original #5096 bug where fork/truncate carried the entire parent
context into the model.

## The remaining edge

`context_messages` is the provider-facing history, so it interleaves `tool_calls` / `tool` rows
**throughout** the transcript, while display `messages` collapses each turn into a single bubble. When
the extra context rows belong to a *later* turn being truncated (rather than a leading prefix), the
positional `prefix + suffix[:keep]` arithmetic can over-keep:

```
ctx = [u1, a1, u2, a2, tool_call, tool_result, a2b]   display = [u1, a1, u2, a2]
keep = 2  (intent: keep only u1, a1)
prefix_len = 3 -> prefix = [u1, a1, u2];  suffix[:2] = [a2, tool_call]
result = [u1, a1, u2, a2, tool_call]   <- the turn-2 you meant to delete partly survives
```

So a mid-history rewind/fork on a **tool-heavy turn** can still leave some post-cut context visible to
the model on the next turn. This is the same *class* as #5096 but a narrower, tool-interleaved case.

**Not a regression** — current `master` (before the bundle) fork-copies the *entire* parent context,
which is strictly worse. This issue tracks closing the remaining gap.

## Why it wasn't fixed in the bundle

The natural robust fix is to align by a **timestamp/id watermark** instead of by position
(`_truncation_watermark_for` already exists and is used for the display side). But an empirical check of
real sessions shows `context_messages` rows **do not carry timestamps** — they're provider-shaped
(`{role, content, tool_calls, tool_call_id}` only), so there's no per-row watermark to filter on without
a larger change to how context rows are persisted/shaped. That's out of scope for the targeted #5096
fixes, which ship the strict improvement now.

## Possible directions

- Persist a stable per-row marker (timestamp or turn-id) on `context_messages` rows so truncation can
  filter by watermark rather than position (the robust fix; touches the streaming write path).
- Or: make the helper turn-aware — walk `context_messages` and cut at the boundary of the last kept
  *display* turn (group tool rows with their owning assistant turn) rather than slicing a flat tail.

## Acceptance

- Mid-history truncate/fork on a tool-interleaved turn leaves **no** post-cut context rows in
  `context_messages`.
- Regression test with an interleaved-tool transcript (and one for `len(ctx) < len(msgs)`,
  post-compaction summary collapse, currently untested).

Refs #5096, #5124, #5126.

### PR DESCRIPTION
**Pull request by rodboev**
## Thinking Path

- The `#5096` context-truncation bundle fixed the broad parent-context leak by routing branch, edit, regenerate, and truncate rewinds through `truncate_context_for_display_keep()`.
- The remaining edge is narrower: provider-facing `context_messages` can contain tool calls, tool results, and assistant continuation rows between display rows, so a positional length-delta cut can keep part of the first unkept turn.
- The fix stays inside the existing helper owner and cuts at the first unkept display anchor instead of treating every extra context row as a leading compaction prefix.

## What Changed

- `api/session_ops.py`: align display rows inside `context_messages` and cut at the first unkept display anchor, while preserving same-length, leading-prefix, and compact-short-context fallback behavior.
- `tests/test_issue_branch_context_at_fork.py`: add regressions for interleaved later tool rows, kept-turn tool rows, and the `len(ctx) < len(msgs)` compact-summary fallback.

## Why It Matters

Mid-history branch, edit, regenerate, and truncate actions should leave the model context at the same boundary the user sees in the display transcript. This prevents post-cut tool rows from leaking into the next model call while keeping the `#5096` improvements intact.

## Verification

- `pytest tests/test_issue_branch_context_at_fork.py -v --timeout=60`
- Full-suite CI context: `pytest tests/ -v --timeout=60`

## Upstream

Closes #5134.

## Model Used

GPT 5.5 via Codex CLI

### CODE REVIEW

File: api/session_ops.py [hunk 1]
```diff
@@ -81,9 +81,63 @@ def truncate_context_for_display_keep(
         return []
     if len(ctx) == len(msgs):
         return ctx[:keep]
+    if len(ctx) < len(msgs):
+        return ctx[:keep]
     if len(msgs) == 0:
```
[the remark this answers is not in the document]
Comment (rodboev) [reply api/session_ops.py]: Fixed in dac21e76. Collapsed the duplicated equality and shorter-context guards into `if len(ctx) <= len(msgs):` at `api/session_ops.py:82`, so the early return keeps the same behavior without the extra branch.

File: api/session_ops.py [hunk 2]
```diff
@@ -81,9 +81,63 @@ def truncate_context_for_display_keep(
...
+    # Cut at the first unkept display turn, or fallback to the last kept turn
+    # if the boundary is not directly alignable.
+    if keep < len(msgs):
+        first_unkept = matches[keep]
+        if first_unkept is not None:
+            return ctx[:first_unkept]
+        if keep > 0:
+            last_kept = matches[keep - 1]
+            if last_kept is not None:
+                return ctx[:last_kept + 1]
```
[the remark this answers is not in the document]
Comment (rodboev) [reply api/session_ops.py:158]: Fixed in dac21e76. When the last kept display row is a user turn, `truncate_context_for_display_keep()` now cuts at the matched kept row instead of carrying tool rows from the discarded assistant turn, and `tests/test_issue_branch_context_at_fork.py:86` covers that boundary explicitly.

**Comment by nesquena-hermes** [comment]
# 🔬 Gate certification — RED ⛔ (good fix, but one alignment edge drops a kept turn — reproduced)

**Certified head:** `sha:dac21e762d11` (rebased `b47bd4e0`) · **PR:** #5212 · _rodboev — "fix(session): drop interleaved tool rows after mid-history truncate (#5134)"_ · **crit=3 data-integrity**
**Verdict:** The fix correctly resolves #5134's over-keep and degrades safely to #5096 — but it introduces a **reproduced regression**: when the model-facing context has a duplicate same-role/same-content row with **no `id`/`timestamp`** to disambiguate, a user-boundary truncate can bind the display row to the *earlier* context-only duplicate and **drop the actually-kept user turn** from `context_messages`. Codex caught it; Opus's APPROVE missed it; **I reproduced it directly against master.** One alignment-edge fix away from green.

### What I ran (isolated worktree `/tmp/wt-gate-5212`, rebased onto current master)
| Gate | Result |
|------|--------|
| Codex (reproduce) | **SHIP ONLY WITH FIXES** — CORE: kept-user-turn dropped on duplicate no-id rows at user boundary |
| Opus (full review) | **APPROVE** — judged the weak fallback "safe"; did not construct the no-id-identical case |
| Full pytest suite (`-p no:xdist`, rebased) | **11088 passed, 0 failed** (green — but does NOT cover this case) |
| PR's own tests | **6 passed** (but all rows carry `id`+`timestamp`, so they never exercise the bare role+content path that fails) |
| My adjudication (reproduced the split) | **Codex is right** — see below |

### ⛔ Blocking — kept user turn dropped (I reproduced master-vs-PR divergence)
`_row_matches()` (api/session_ops.py:93) falls through to a weak `role`+`content` match when `id`/`timestamp` are absent or conflict. With two byte-identical no-id user rows (a context-only duplicate before the real kept row), the greedy forward alignment binds the display row to the *first* (hidden) duplicate, so `last_kept` points at the wrong row, and the user-boundary cut `ctx[:last_kept+1]` (line 140) **excludes the real kept user turn**. Reproduced exactly:
```
ctx  = [user "u1"(no id, hidden dup), user "u1"(no id, real), assistant "a1"]
disp = [user "u1", assistant "a1"];  keep=1
  origin/master → ['u1','u1']   (keeps the span through the real row)
  this PR       → ['u1']        (drops the real kept user turn from context_messages)
```
Display/`session.messages` still shows the turn, so it's a **model-facing context-fidelity regression after fork/truncate**, not data loss or a crash. **Severity: real CORE-class, moderate likelihood** — it needs identical-content adjacent user rows with no id/timestamp (Opus notes production provider rows "often have neither," and `_append_recovered_turn_to_context` even strips timestamp), though byte-identical adjacent user turns are uncommon.

**Why the green suite + PR tests missed it:** the PR's 4 tests give every row an `id` and `timestamp`, so they only exercise the strong-match branches — never the bare `role`+`content` fallback that fails. (Opus flagged this test-fidelity gap as a nit; it's actually the bug's trigger.)

**Fix-spec (Codex, sound):** in `_row_matches`, make conflicting `id`/`timestamp` **fail closed** (don't fall through to plain role+content), include tool metadata in the comparison, and only use the weak fallback when the match is unambiguous in the remaining context. Add a regression test with **duplicate same-content user rows and no id/timestamp** at a user-boundary cut.

### Adjudication note (advisor split — the layer's core value)
Codex (SHIP-WITH-FIXES) and Opus (APPROVE) directly disagreed on whether the weak fallback is safe. I reproduced the exact case: Codex is correct. Opus's "cannot leak post-cut rows" reasoning holds for id-bearing rows (the forward `next_ctx_idx` advance prevents back-binding) but **not** for bare identical role+content rows where the earlier duplicate is the first forward match. Don't ship on the APPROVE alone.

### What's good (keep it)
- ✅ Correctly resolves #5134 (interleaved tool tail dropped after the cut); my own tests confirmed kept-turn tool rows are preserved and post-cut rows dropped for id-bearing transcripts.
- ✅ Degrades to the #5096 leading-prefix fallback when alignment is unreliable; `len(msgs)==0` branch still reachable (Greptile's "unreachable" claim is wrong — ignore it).
- ✅ Pure helper, no shared-state mutation; rebases cleanly onto current master.

### Recommendation to the next agent
**Bounce for the `_row_matches` fail-closed fix + a no-id-duplicate regression test, then re-gate.** This is a good, surgical fix that's one alignment edge short — the kept-user-drop is a reproduced regression vs master in the duplicate-no-id case, not a theoretical one. Fix is well-defined and small. I applied `gate-fail`; the regression is objective and reproduced. Cert valid only at `sha:dac21e762d11`.

---
_Gate-certifier layer (warm-up → **gate** → release). I do not merge/tag/deploy/close. Cert valid only at `sha:dac21e762d11`; a new push invalidates it → re-gate._

**Comment by nesquena-hermes** [comment]
Bouncing for one alignment-edge fix — the gate certification above (sha `dac21e762d11`) reproduced a real **CORE-class regression vs master**, so this needs a small rework before it can ship.

**The good part (keep it):** the fix correctly resolves #5134 (interleaved tool tail dropped after a mid-history truncate) and degrades to the #5096 leading-prefix fallback. Pure helper, no shared-state mutation, rebases clean.

**The blocker (reproduced, not theoretical):** `_row_matches()` falls through to a weak `role`+`content` match when `id`/`timestamp` are absent. With two byte-identical no-id user rows (a context-only duplicate before the real kept row), the greedy forward alignment binds the display row to the *earlier hidden* duplicate, so the user-boundary cut `ctx[:last_kept+1]` **drops the actually-kept user turn** from `context_messages`:
```
ctx  = [user "u1"(no id, hidden dup), user "u1"(no id, real), assistant "a1"]
disp = [user "u1", assistant "a1"];  keep=1
  origin/master → ['u1','u1']   (keeps through the real row)
  this PR       → ['u1']        (drops the real kept user turn)
```
Display still shows the turn, so it's a model-facing **context-fidelity** regression after fork/truncate (not data loss/crash), but it's real — provider rows often carry neither id nor timestamp, and `_append_recovered_turn_to_context` strips timestamp. Your 4 tests give every row an `id`+`timestamp`, so they only exercise the strong-match branches and never hit this path.

**Fix-spec (small, well-defined):**
1. In `_row_matches`, make conflicting `id`/`timestamp` **fail closed** — don't fall through to plain role+content.
2. Include tool metadata in the comparison.
3. Only use the weak role+content fallback when the match is unambiguous in the remaining context.
4. Add a regression test with **duplicate same-content user rows and no id/timestamp** at a user-boundary cut (the case the current tests miss).

Re-gate on re-push and this should land — it's genuinely one edge short. Thanks @rodboev.

**Comment by rodboev** [comment]
Reworked the remaining alignment edge on head `4af8c36b`. `api/session_ops.py` now treats conflicting `id` or `timestamp` data as a failed match, includes tool metadata in the row signature, and only accepts the weak role/content fallback when the remaining context has a single candidate. I also added the duplicate no-id/no-timestamp user-boundary regression in `tests/test_issue_branch_context_at_fork.py`, so the kept user turn stays in `context_messages` instead of binding to the earlier hidden duplicate.

**Comment by nesquena-hermes** [comment]
Re-gated the convergence — the no-id-duplicate kept-user-drop I bounced for is **fixed** (the `weak_matches > 1 → None` fail-closed + the new `_keeps_real_user_turn_when_duplicate_rows_lack_identity` test both check out). Thanks for the rework. But the re-gate (Codex, reproduced vs merge-base) surfaced one **new CORE edge the fail-closed change introduced** — close, but one more turn needed.

**The remaining issue (reproduced):** when the **first-unkept anchor is ambiguous** (multiple weak matches → `_first_match_from` returns `None`), the caller falls through to `return ctx[:last_kept + 1]`. If the last *kept* display row is a **non-user turn with tool rows after it** (an assistant `tool_calls` turn followed by its tool result), `last_kept + 1` cuts BEFORE those kept-turn tool rows — dropping the interleaved tool tail that #5134 set out to preserve. Verified: for a kept assistant tool-call turn with a following tool result and a duplicated no-id/no-timestamp unkept user anchor, PR head returns `['user:u1','assistant:a1']` while merge-base preserved `tool:kept-tool-result`.

So the fail-closed fix traded the kept-*user*-drop for a kept-*tool*-drop in the ambiguous-first-unkept case.

**Fix-spec (Codex, sound):**
1. `api/session_ops.py` (~130-159): have `_first_match_from()` carry back the **earliest ambiguous weak-match index** (don't just return `None`). When `first_unkept` is ambiguous, cut at that earliest candidate for **non-user** kept rows (so the kept turn's trailing tool rows are preserved), while keeping the existing `last_kept + 1` user-boundary strip for the user case.
2. Add a regression: `keep=2`, kept assistant has `tool_calls`, a tool result follows it, and the unkept user anchor is duplicated with no id/timestamp → expected context `[u1, a1, kept-tool-result]` (no duplicated unkept user, no dropped tool result).

The user-boundary case is correct now; this is just the symmetric non-user (tool-tail) case in the ambiguous branch. Re-gate on re-push. Thanks @rodboev — genuinely close.

**Comment by rodboev** [comment]
Reworked the ambiguous-anchor boundary on head `4af8c36b`'s successor. `_first_match_from()` now reports the earliest ambiguous weak-match candidate instead of flattening it into plain no-match, and `truncate_context_for_display_keep()` uses that candidate for non-user kept rows. That preserves the kept assistant turn's trailing tool result while still stripping duplicated unkept user rows.

I added the regression you described: kept assistant with `tool_calls`, a following kept tool result, then a duplicated no-id/no-timestamp unkept user anchor. The expected context is now `u1`, `a1`, and the kept tool result only.

**Comment by nesquena-hermes** [comment]
Shipped in **v0.51.760** (via #5249). Thanks @rodboev — converged cleanly over two gate rounds: the truncate matcher now fails closed on ambiguous (id-less duplicate) matches, preserving both the kept user turn AND the kept assistant turn's trailing tool rows. Codex SAFE, full suite green. Closes #5134-class truncate data loss.
